lib/veri_policy.ml

```diff
@@ -0,0 +1,230 @@
+
+    type t = {src : V.t; dst : V.t} [@@deriving fields]
+    let make src dst = {src; dst}
+    let label _ = ()
+  end
+  type dir = Succ | Pred
+
+  let iter dir f (expect, workers, jobs) v = 
+    match v,dir with
+    | V.Source,Pred -> ()
+    | V.Source,Succ ->
+      Array.iteri workers ~f:(fun i _  ->
+          f @@ E.make V.Source (V.Person i))
+    | V.Sink,Pred ->
+      Array.iteri jobs ~f:(fun i _ ->
+          f @@ E.make (V.Person i) V.Sink)
+    | V.Sink,Succ -> ()
+    | V.Person i as p,Pred ->
+      f @@ E.make V.Source p
+    | V.Person i as p,Succ ->
+      Array.iteri jobs ~f:(fun j job ->
+          if sat_events expect workers.(i) job
+          then f (E.make p (V.Task j)))
+    | V.Task j as t,Succ ->
+      f @@ E.make t V.Sink
+    | V.Task j as t,Pred ->
+      Array.iteri workers ~f:(fun i worker ->
+          if sat_events expect worker jobs.(j)
+          then f (E.make (V.Person i) t))
+
+  let iter_succ_e = iter Succ
+  let iter_pred_e = iter Pred
+end
+
+module F = struct
+  type t = int
+  type label = unit
+  let max_capacity () = 1
+  let min_capacity () = 0
+  let flow () = min_capacity ()
+  let add = (+)
+  let sub = (-)
+  let zero = 0
+  let compare = Int.compare
+end
+
+module FFMF = Flow.Ford_Fulkerson(G)(F)
+
+let single_match trial events =
+  List.filter ~f:(sat_event trial) (Set.to_list events)
+
+let match_right t events = 
+  match single_match t.right_trial events with 
+  | [] -> None
+  | ms -> Some (Right ms)  
+
+let match_left t events = 
+  match single_match t.left_trial events with 
+  | [] -> None
+  | ms -> Some (Left ms)  
+
+let match_both t left right = 
+  let workers = Set.to_array left in      
+  let jobs = Set.to_array right in
+  let (flow,_) = FFMF.maxflow (t.both_trial, workers, jobs) G.V.Source G.V.Sink  in
+  Array.foldi workers ~init:[] 
+    ~f:(fun i acc w ->   
+        match Array.findi jobs ~f:(fun j e ->
+            flow (G.E.make (G.V.Person i) (G.V.Task j)) <> 0) with
+        | None -> acc 
+        | Some (_,e) -> (w, e) :: acc) |> function 
+  | [] -> None
+  | ms -> Some (Both ms)
+
+let is_empty_insn  t = Field.is_empty (Rule.insn t.rule)
+let is_empty_left  t = Field.is_empty (Rule.left t.rule)
+let is_empty_right t = Field.is_empty (Rule.right t.rule)
+let is_sat_insn t insn = not (is_empty_insn t) && sat t.insn_trial insn
+
+let match_events' t insn events events' =
+  match is_sat_insn t insn with
+  | false -> None
+  | true -> 
+    let left = Set.diff events events' in
+    let right = Set.diff events' events in
+    match is_empty_left t, is_empty_right t with
+    | true, _ -> match_right t right
+    | _, true -> match_left t left
+    | _ -> match_both t left right
+
+let match_events rule insn events events' =
+  match_events' (make_entry rule) insn events events'
+
+let remove what from = 
+  let not_exists e = not (List.exists what ~f:(fun e' -> e = e')) in
+  Set.filter ~f:not_exists from 
+
+let remove_matched events events' = function
```

Comment:
according to this code, I'm understanding your match type correctly. It indeed should be just a pair of lists.